File: 17-06-2025_Count_the_number_of_arrays_with_k_matching_adjacent_elements.java
class Solution {
    static final int MOD = 1_000_000_007;
    long[] fact, invFact;

    long modPow(long base, long exp) {
        long result = 1;
        base %= MOD;
        while (exp > 0) {
            if ((exp & 1) == 1) result = (result * base) % MOD;
            base = (base * base) % MOD;
            exp >>= 1;
        }
        return result;
    }

    void precomputeFactorials(int n) {
        fact = new long[n + 1];
        invFact = new long[n + 1];
        fact[0] = 1;
        for (int i = 1; i <= n; i++) {
            fact[i] = (fact[i - 1] * i) % MOD;
        }
        invFact[n] = modPow(fact[n], MOD - 2);
        for (int i = n - 1; i >= 0; i--) {
            invFact[i] = (invFact[i + 1] * (i + 1)) % MOD;
        }
    }

    long comb(int n, int k) {
        if (k < 0 || k > n) return 0;
        return (((fact[n] * invFact[k]) % MOD) * invFact[n - k]) % MOD;
    }

    public int countGoodArrays(int n, int m, int k) {
        precomputeFactorials(n);
        long ways = comb(n - 1, k) * modPow(m - 1, n - 1 - k) % MOD;
        ways = ways * m % MOD;
        return (int) ways;
    }
}

File: 17-06-2025_Count_the_number_of_arrays_with_k_matching_adjacent_elements.py
class Solution:
    MOD = 10**9 + 7

    def mod_pow(self, base: int, exp: int) -> int:
        result = 1
        while exp > 0:
            if exp % 2 == 1:
                result = result * base % self.MOD
            base = base * base % self.MOD
            exp //= 2
        return result

    def precompute_factorials(self, n: int):
        self.fact = [1] * (n + 1)
        self.inv_fact = [1] * (n + 1)
        for i in range(1, n + 1):
            self.fact[i] = self.fact[i - 1] * i % self.MOD
        self.inv_fact[n] = self.mod_pow(self.fact[n], self.MOD - 2)
        for i in range(n - 1, -1, -1):
            self.inv_fact[i] = self.inv_fact[i + 1] * (i + 1) % self.MOD

    def comb(self, n: int, k: int) -> int:
        if k < 0 or k > n:
            return 0
        return self.fact[n] * self.inv_fact[k] % self.MOD * self.inv_fact[n - k] % self.MOD

    def countGoodArrays(self, n: int, m: int, k: int) -> int:
        self.precompute_factorials(n)
        ways = self.comb(n - 1, k) * self.mod_pow(m - 1, n - 1 - k) % self.MOD
        ways = ways * m % self.MOD
        return ways
